Adjust bootloader argument rules to work in bootable containers

jan-cerny opened the pull request:
This commit extends templates `grub2_bootloader_argument` and  `grub2_bootloader_argument_absent` for bootable containers.

The rules that use these templates will be marked with the `system_with_kernel` platform to extend their applicability also to bootable containers.

The bootable containers use a special mechanism for specifying the kernel arguments. In bootable containers, the kernel build arguments are specified in TOML files in special directory  `/usr/lib/bootc/kargs.d/`.  For more details, see:
https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/using_image_mode_for_rhel_to_build_deploy_and_manage_operating_systems/managing-kernel-arguments-in-bootc-systems#how-to-add-support-to-inject-kernel-arguments-with-bootc_managing-kernel-arguments-in-bootc-systems
Based on that, the OVAL will check the `/usr/lib/bootc/kargs.d/` and the remediation will remediate this directory if a bootable container image is built.

To support this effort, this PR introduces new product property `bootable_containers_supported`. By default, this variable is set to `"false"` and RHEL 9 and 10 products set this variable to `"true"`.

jan-cerny commented on shared/templates/grub2_bootloader_argument/oval.template:
```diff
@@ -33,10 +34,14 @@
 {{% set system_with_bios_and_uefi_support = true %}}
 {{%- endif -%}}
 
+{{% if product in ["rhel9", "rhel10"] -%}}
+{{% set bootable_containers_supported = true %}}
+{{%- endif -%}}
```
I think this increases technical debt.

matusmarhefka replied:
Yes, so can we go without it? Just have a separate criteria which would check bootc environment + argument existance/non-existance in the `/usr/lib/bootc/kargs.d/*.toml` files?

matusmarhefka commented on shared/templates/grub2_bootloader_argument_absent/bash.template:
```diff
@@ -3,4 +3,8 @@
    Product-specific categorization should be synced across all template content types
 -#}}
+if {{{ bash_bootc_build() }}} ; then
+    sed -i -E '/kargs\s+=\s+\["{{{ ARG_NAME }}}.*"\]/d' /usr/lib/bootc/kargs.d/*.toml
```
Based on https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/using_image_mode_for_rhel_to_build_deploy_and_manage_operating_systems/managing-kernel-arguments-in-bootc-systems#how-to-add-support-to-inject-kernel-arguments-with-bootc_managing-kernel-arguments-in-bootc-systems there can be multiple kernel arguments inside a single `kargs` line (e.g. `kargs = ["mitigations=auto,nosmt"]`) which is not handled in our remediation here. Our current remediation would even remove these remaining options if there would be the `kargs` line starting with `{{{ ARG_NAME }}}`. We also need to handle possibility of multiple kernel arguments in a single `kargs` line in our OVAL.

One more thing, the `kargs` line can also be followed by the `match-architectures` line so I believe we need to remove the `match-architectures` in case we remove the whole `kargs` line.

dodys reviewed (approved):
lgtm, thanks

matusmarhefka commented on shared/templates/grub2_bootloader_argument_absent/bash.template:
```diff
@@ -3,4 +3,9 @@
    Product-specific categorization should be synced across all template content types
 -#}}
+if {{{ bash_bootc_build() }}} ; then
+    sed -i -E "/kargs\s*=\s*\[\s*\"{{{ ARG_NAME }}}=[^\"]*\"\s*]/d" "$KARGS_DIR/*.toml"
```
I would update the first line so it also deletes following `match-architectures` line if it is found (and also count with empty lines between `kargs` and `match-architectures`:
```
sed -i -E "/kargs\s*=\s*\[\s*\"{{{ ARG_NAME }}}=[^\"]*\"\s*]/{:a;N;/^\n$/ba;N;/match-architectures.*/d;}" "$KARGS_DIR/*.toml"
```
The second sed command should take care of just removing the specified kernel argument among other arguments, but as it is now it won't remove the comma character if our argument is followed by another argument or if it is preceded by another argument. Therefore, I would update the command like this:
```
sed -i -E -e "s/^(\s*kargs\s*=\s*\[.*)\"{{{ ARG_NAME }}}=[^\"]*\"[,[:space:]]*(.*]\s*)/\1\2/" -e "s/^(\s*kargs.*),[[:space:]]*\]$/\1\]/" "$KARGS_DIR/*.toml"
```

jan-cerny commented:
I have changed the Bash remediation according to your suggestion. Then, I had changed platform expressions in the whole group to simplify and unify the applicability. I have also added the explanation of bootc kernel arguments to the rule description.

jan-cerny commented:
I have improved regular expression so that it will match kernel arguments
presents in lists that have more than 1 item.

jan-cerny commented:
/packit build

teacup-on-rockingchair reviewed (approved):
LGTM 👍

Xeicker commented on shared/templates/grub2_bootloader_argument_absent/oval.template:
```diff
@@ -32,6 +32,7 @@
 <def-group>
   <definition class="compliance" id="{{{ _RULE_ID }}}" version="2">
     {{{ oval_metadata("Ensure " + ARG_NAME + " is not set in the kernel line in /etc/default/grub.") }}}
+    <criteria operator="OR">
```
shouldn't this be "AND"?

I'm thinking that if the system supports bootable containers and it doesn't contain the argument in `/usr/lib/bootc/kargs.d/*.toml` it will pass even if the argument is present somewhere else

jan-cerny replied:
Good point! I will try to adjust this OVAL to prevent that situation.

jan-cerny commented:
Thanks everyone for approving.
